libs/cloud-database/src/database.rs

```diff
@@ -70,33 +75,84 @@ impl CloudDatabase {
     #[instrument(skip(self))]
     pub async fn user_login(&self, login: UserLogin) -> Result<Option<UsersModel>, DbErr> {
         info!("database user_login enter");
-        Users::find()
+        let Ok(user) = Users::find()
+            .column(UsersColumn::Password)
             .filter(UsersColumn::Email.eq(login.email))
-            .filter(UsersColumn::Password.eq(login.password))
             .one(&self.pool)
-            .await
+            .await else {
+                return Err(DbErr::RecordNotFound(
+                    "Failed to find the user".to_string(),
+                ));
+            };
+
+        let hashed_password = user
+            .clone()
+            .ok_or(DbErr::RecordNotFound("Failed to find the user".to_string()))?
+            .password
+            .ok_or(DbErr::RecordNotFound("Failed to find the user".to_string()))?;
+
+        let parsed_hash = PasswordHash::new(&hashed_password).or(Err(DbErr::RecordNotFound(
+            "Failed to find the user".to_string(),
+        )))?;
+
+        if Argon2::default()
+            .verify_password(login.password.as_bytes(), &parsed_hash)
+            .is_err()
+        {
+            return Err(DbErr::RecordNotFound("Failed to find the user".to_string()));
+        }
+
+        Ok(user)
     }
 
     #[instrument(skip(self, token))]
     pub async fn refresh_token(&self, token: RefreshToken) -> Result<Option<UsersModel>, DbErr> {
         info!("database refresh_token enter");
-        Users::find()
+        let Ok(user) = Users::find()
+            .column(UsersColumn::TokenNonce)
             .filter(UsersColumn::Id.eq(token.user_id))
-            .filter(UsersColumn::TokenNonce.eq(token.token_nonce))
             .one(&self.pool)
-            .await
+            .await else {
+                return Err(DbErr::RecordNotFound(
+                    "Failed to find the user".to_string(),
+                ));
+            };
+
+        let token_nonce = user
+            .clone()
+            .ok_or(DbErr::RecordNotFound("Failed to find the user".to_string()))?
+            .token_nonce
+            .ok_or(DbErr::RecordNotFound("Failed to find the user".to_string()))?;
+
+        if !constant_time_eq(&token_nonce.to_le_bytes(), &token.token_nonce.to_le_bytes()) {
```

Comment:
What are the considerations for replacing In-Database filter with self-comparison? In the cloud version, we use pg as the database. Experience has shown that in most cases, the performance consumed by fetching data locally is much greater than the performance improvement brought by self-comparison.

Reply:
Prevent [Timing Attack](https://en.wikipedia.org/wiki/Timing_attack) as the `token_nonce` is only i16 long.
https://security.stackexchange.com/questions/162211/is-sql-where-queries-in-authentication-prone-to-timing-attacks-if-one-or-no-dat

Reply:
Also, `token_nonce`'s i16 is too short, and itself can be a vulnerability.

Reply:
Sorry, I just find that it is aes encrypted, no issue now.
I will change it back.